src/dataflow/src/decode/regex.rs

```diff
@@ -24,62 +24,73 @@ pub fn regex<G>(
+                    position: line_no,
+                    upstream_time_millis: _,
+                } in &*lines
+                {
+                    let line = match str::from_utf8(&line) {
+                        Ok(line) => line,
+                        Err(_) => {
+                            let line_no = match line_no {
+                                Some(line_no) => line_no.to_string(),
+                                None => "unknown".into(),
+                            };
+                            let line_prefix = String::from_utf8_lossy(&line)
+                                .chars()
+                                .take(64)
+                                .collect::<String>();
+                            session.give((
+                                Err(DataflowError::DecodeError(DecodeError::Text(format!(
+                                    "Regex error in source {} at lineno {}: invalid UTF-8, line prefix: {:?}",
+                                    name, line_no, line_prefix
+                                )))),
+                                *cap.time(),
+                                1,
+                            ));
+                            continue;
+                        }
+                    };
 
-                        let captures = match regex.captures(line) {
-                            Some(captures) => captures,
-                            None => continue,
-                        };
+                    let captures = match regex.captures(line) {
+                        Some(captures) => captures,
+                        None => continue,
+                    };
 
-                        // Skip the 0th capture, which is the entire match, so that
-                        // we only output the actual capture groups.
-                        let datums = captures
-                            .iter()
-                            .skip(1)
-                            .map(|c| Datum::from(c.map(|c| c.as_str())))
-                            .chain(iter::once(Datum::from(*line_no)));
+                    // Skip the 0th capture, which is the entire match, so that
+                    // we only output the actual capture groups.
+                    let datums = captures
+                        .iter()
+                        .skip(1)
+                        .map(|c| Datum::from(c.map(|c| c.as_str())))
+                        .chain(iter::once(Datum::from(*line_no)));
 
-                        session.give((row_packer.pack(datums), *cap.time(), 1));
-                    }
-                });
-            }
-        })
-        .as_collection()
+                    session.give((Ok(row_packer.pack(datums)), *cap.time(), 1));
+                }
+            });
+        }
+    });
+
+    let (oks, errs) = stream.ok_err(|(data, time, diff)| match data {
+        Ok(data) => Ok((data, time, diff)),
+        Err(err) => Err((err, time, diff)),
+    });
```

Comment:
@frankmcsherry don't we have a utility function somewhere that does this?

Reply:
It's possible that `map_fallible(|x| x)` does this. idk!

Reply:
I guess not, but `stream.as_collection().map_fallible(|x| x)` should do it.

That might be less clear than the present code, though.

Reply:
Ah, we commented the same thing at the same time :D

Reply:
I like the idea of reducing code here. I cannot return the output from `map_fallible` directly, because the error stream needs to be wrapped in an `Option`. Once we finish with this error handling, we should consider requiring a collection instead of an option for the error stream and then we can return the results from map_fallible directly.